### PR TITLE
Integrate cp with s3transfer

### PR DESCRIPTION
**Pull request by kyleknap**
This is the last pull request in integrating s3transfer with ``cp`` and ``cp --recursive``.

All of the test pass with minimal changes to the existing test suite so that is good. If you want to know why a particular test may have changed, I would recommend checking out the commit. I tried to be detailed on why I changed particular tests.

While this PR is smaller than some of my previous PR's, this is probably a good one to start trying to address any design feedback or ideas of the complete port of cp and how it affects ports of future commands.

For example, two larger design questions that I would like some conversation on include:

* How to manage submitters? Should these be injected into a general S3Handler? Should they be built into the S3Handler and you create various S3Handler subclasses? Or should there be one S3TransferHandler class with all of the submitters built in (which I chose because it is simplest given what is currently supported)?

* Thoughts on introducing builders for the various components in the future? For example having an S3TransferHandler builder and maybe eventually a FileGenerator, Filter, etc builders so that not all of this logic is not under CommandArchitecture. I only do that right now because the builder abstraction is not really needed as there is not too many different parameters that needs to be taken into account, but there may be in the future (you can already see it sort of forming with the result printer).

Let me know what you think.

cc @jamesls @JordonPhillips

### CODE REVIEW

File: awscli/customizations/s3/subcommands.py [hunk 1]
```diff
@@ -940,6 +952,10 @@ def run(self):
         s3_stream_handler = S3TransferStreamHandler(
             self.session, self.parameters, result_queue=result_queue)
 
+        s3_transfer_hanlder = s3handler
```
Comment (jamesls) [inline awscli/customizations/s3/subcommands.py]: typo, hanlder

File: awscli/customizations/s3/subcommands.py [hunk 2]
```diff
@@ -1008,6 +1024,37 @@ def run(self):
             rc = 2
         return rc
 
+    def _get_s3_transfer_handler(self, result_queue):
```
Comment (jamesls) [inline awscli/customizations/s3/subcommands.py]: Seems like it's worth pulling this into a separate class, both for testability, but also as a clean slate to start integrating more and more of the s3transfer logic over.

File: awscli/customizations/s3/s3handler.py [hunk 3]
```diff
@@ -523,6 +525,80 @@ def _process_transfer(self, future):
...
+            )
+
+    def _shutdown(self):
+        self._result_queue.put(ShutdownThreadRequest())
+        self._result_processor.join()
+
+    def _enqueue(self, fileinfos):
+        for fileinfo in fileinfos:
+            for submitter in self._submitters:
+                if submitter.is_valid_submission(fileinfo):
```
Comment (jamesls) [inline awscli/customizations/s3/s3handler.py]: Seems like this isn't really about a "valid" submission.  It's more if the particular submitter understands the request / knows how to submit it.  What about names like `can_submit`, `can_handle`, etc instead of `is_valid_submission`?

File: awscli/customizations/s3/s3handler.py [hunk 4]
```diff
@@ -523,6 +525,80 @@ def _process_transfer(self, future):
...
+
+        :type result_recorder: ResultRecorder
+        :param result_recorder: Result recorder to get status of transfer
+
+        :type result_processor: ResultProcessor
+        :param result_processor: Result processor to process status updates
+            of transfer
+        """
+        self._transfer_manager = transfer_manager
+        self._result_queue = result_queue
```
Comment (jamesls) [inline awscli/customizations/s3/s3handler.py]: Seems like there's a missing abstraction here.  You have a result queue, recorder, and processor, but at this level of abstraction, it seems too low level.  Especially in methods like `_shutdown()` where you have to coordinate how to shutdown the "result subsystem" by first enqueueing a shutdown request, then joining on a thread.  Seems like this can be packaged behind a class.

**Comment by jamesls** [comment]
High level seems fine.  Still not entirely sold on the idea of iterating through every type of submitter for every single file we transfer, but I would agree that in its current form with only 3 submitters it's not really that big of a deal.

File: awscli/customizations/s3/subcommands.py (hunk 2 above)
Comment (kyleknap) [inline awscli/customizations/s3/subcommands.py]: Cool. I was thinking that an abstraction was eventually going to be needed. Just was not sure when, but I'm totally with creating the abstraction now.

File: awscli/customizations/s3/s3handler.py (hunk 3 above)
Comment (kyleknap) [inline awscli/customizations/s3/s3handler.py]: ``can_submit()`` is a better name I will do that.

File: awscli/customizations/s3/s3handler.py (hunk 4 above)
Comment (kyleknap) [inline awscli/customizations/s3/s3handler.py]: Yeah it felt a little weird that I was passing in a result recorder and a result processor that already had been built with a result recorder to the initializer. How does an interface like this sound?
```py
class ResultProcessorManager:
    def __init__(self, result_queue, result_recorder, result_processor):
        self.result_queue = result_queue
    def start(self):
    def shutdown(self):
    def get_final_result(self):
    def __enter__(self):
    def __exit__(self, exc_type, *args):
```
That way all I would have to pass in is this result processor manager to the s3 transfer handler. If I do this, should I make it a separate PR? I feel like that may be nicer.

Comment (jamesls) [inline awscli/customizations/s3/s3handler.py]: Wouldn't you need a method to submit a result for processing if your'e no longer passing a `result_queue` to this this class?

Separate PR seems fine.

Comment (kyleknap) [inline awscli/customizations/s3/s3handler.py]: Possibly. So this is what I am planning. Let me know if that makes sense.

1) Change the interface of ``S3TransferHandler`` to:
```py
class S3TransferHandler(object):
    def __init__(self, transfer_manager, cli_params, result_queue):
```
2) Have the ``CommandResultRecorder`` live outside of ``S3TransferHandler`` because it eclipses more than the ``S3TransferHandler`` it covers the ``FileGenerators`` and any of the other abstractions. So I was going to have it do something like this:
```py
with command_result_recorder:
    fileinfos = filegenerator.call()
    s3_transfer_handler.call(fileinfos)
return command_result_recorder.get_command_result()
```

Comment (jamesls) [inline awscli/customizations/s3/s3handler.py]: Interesting.  I was going down the path of abstracting away results so that code doesn't have to deal directly with `result_queue`, which is a bit of an implementation detail.  Something like:

```
with command_result_recorder as r:
    fileinfos = ...
    s3_transfer_handler.call(fileinfos, r)

# In s3 transfer handler
def call(..., reporter):
    try:
       # ...
    except Exception as e:
        reporter.report_result(ErrorResult(exception=e))
```

Sounds like what you're shooting for is to just abstract away the stop/start/final_result part of results and keep submission as is. I think that's an improvement over what's there now, but I'm concerned that separating out the lifecycle of the reporting (start/stop) from the mechanism for reporting (result_queue) could lead to subtle bugs, e.g submission to results queue after result thread is shutdown.

**Comment by kyleknap** [comment]
So I updated this PR based on the PR for CommandResultRecorder. So it would be good if https://github.com/aws/aws-cli/pull/2139 could get reviewed first before completing this one.

File: awscli/customizations/s3/s3handler.py [hunk 5]
```diff
@@ -523,6 +525,118 @@ def _process_transfer(self, future):
...
+            result_queue, result_processor_handlers)
+        command_result_recorder = CommandResultRecorder(
+            result_queue, result_recorder, result_processor)
+
+        return S3TransferHandler(
+            transfer_manager, self._cli_params, command_result_recorder)
+
+    def _add_result_printer(self, result_recorder, result_processor_handlers):
+        result_printer = None
+        if self._cli_params.get('quiet'):
```
Comment (kyleknap) [inline awscli/customizations/s3/s3handler.py]: I wonder if this warrants having a ``ResultPrinterFactory``? It is near impossible to test this without patching or running somewhat of a functional test by seeing what gets printed out when the s3 transfer handler is called. We do have functional tests for this so I do not feel as bad for not having a unit test for this specific logic, but would be nice to unit test it.

**Comment by kyleknap** [comment]
With the referenced PR merged and this one rebased. This should be good to look at again. I believe I have all of the previous feedback addressed.

File: awscli/customizations/s3/s3handler.py (hunk 5 above)
Comment (jamesls) [inline awscli/customizations/s3/s3handler.py]: I think if we add any more result printer instances we should probably go ahead and create a separate `ResultPrinterFactory`.

**Comment by jamesls** [comment]
:shipit:  Looks good to me.

File: awscli/customizations/s3/s3handler.py (hunk 5 above)
Comment (kyleknap) [inline awscli/customizations/s3/s3handler.py]: Sounds like a plan!
